[AArch64] Improve lowering for scalable masked deinterleaving loads

david-arm opened the pull request:
For IR like this:

  %mask = ... @llvm.vector.interleave2(<vscale x 16 x i1> %a, <vscale x 16 x i1> %a)
  %vec = ... @llvm.masked.load(..., <vscale x 32 x i1> %mask, ...)
  %dvec = ... @llvm.vector.deinterleave2(<vscale x 32 x i8> %vec)

where we're deinterleaving a wide masked load of the supported type
and with an interleaved mask we can lower this directly to a ld2b
instruction. Similarly we can also support other variants of ld2
and ld4.

This PR adds a DAG combine to spot such patterns and lower to ld2X
or ld4X variants accordingly, whilst being careful to ensure the
masked load is only used by the deinterleave intrinsic.

c-rhodes reviewed (commented):
thanks for patch Dave, mostly LGTM, but there is a fair few codepaths that are missing tests I've pointed out

c-rhodes reviewed (approved):
thanks for updates, sorry for the noise pointing out all the untested codepaths in the first round of review, it's just there's a ton of constraints in the pattern that's pretty overwhelming so I went through and added unreachables to figure out what was what, and noted ones that didn't fire.

I've got some more suggestions to improve the tests, mostly minor, but the actual code changes LGTM, cheers!

david-arm committed:
Add tests

david-arm committed:
[AArch64] Improve lowering for scalable masked deinterleaving loads

For IR like this:

  %mask = ... @llvm.vector.interleave2(<vscale x 16 x i1> %a, <vscale x 16 x i1> %a)
  %vec = ... @llvm.masked.load(..., <vscale x 32 x i1> %mask, ...)
  %dvec = ... @llvm.vector.deinterleave2(<vscale x 32 x i8> %vec)

where we're deinterleaving a wide masked load of the supported type
and with an interleaved mask we can lower this directly to a ld2b
instruction. Similarly we can also support other variants of ld2
and ld4.

This PR adds a DAG combine to spot such patterns and lower to ld2X
or ld4X variants accordingly, whilst being careful to ensure the
masked load is only used by the deinterleave intrinsic.

c-rhodes commented on llvm/test/CodeGen/AArch64/scalable_masked_deinterleaved_loads.ll:
```diff
@@ -0,0 +1,460 @@
+  ret <vscale x 8 x i16> %add1
+}
+
+define <vscale x 4 x i16> @foo_ld2_nxv8i8_exti16(<vscale x 4 x i1> %mask, ptr %p) {
```
sure no worries, thanks for clarifying 👍

c-rhodes commented:
updates LGTM, cheers!

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -27015,6 +27016,120 @@ performScalarToVectorCombine(SDNode *N, TargetLowering::DAGCombinerInfo &DCI,
+    }
+
+    // Make sure the inputs to the vector interleave are identical.
+    for (unsigned I = 1; I < NumParts; I++) {
+      if (InterleaveOp->getOperand(I) != InterleaveOp->getOperand(0))
+        return SDValue();
+    }
```
```suggestion
    if (!llvm::all_equal(InterleaveOp->op_values()))
      return SDValue();
```

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -27015,6 +27016,120 @@ performScalarToVectorCombine(SDNode *N, TargetLowering::DAGCombinerInfo &DCI,
+
+    for (unsigned I = 0; I < NumParts; I++) {
+      SDValue ConcatOp = Mask->getOperand(I);
+      if (ConcatOp.getResNo() != I ||
+          ConcatOp.getNode() != InterleaveOp.getNode())
```
If you make `InterleaveOp` an SDNode (which is how you're using it anyway for several of the cases by using `->`) you can simplify this to `if (Mask.getOperand(I) == SDValue(InterleaveOp, I))`

david-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -27015,6 +27016,120 @@ performScalarToVectorCombine(SDNode *N, TargetLowering::DAGCombinerInfo &DCI,
+
+    NarrowMask = InterleaveOp->getOperand(0);
+  } else { // ISD::SPLAT_VECTOR
+    auto *SplatVal = dyn_cast<ConstantSDNode>(Mask->getOperand(0));
+    if (!SplatVal || SplatVal->getZExtValue() != 1)
+      return SDValue();
```
I wasn't sure if we should perform the DAG combine if the mask is all-false because we will then generate an intrinsic that we know is a nop and also the DAGCombiner will not understand and not optimise away. I thought that leaving the original code left a greater chance of the whole lot being deleted.

paulwalker-arm replied:
I would expect the all-false case to be caught by instcombine so there would be no need for the code generator to worry about it.

david-arm committed:
Address review comments

david-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -27015,6 +27016,120 @@ performScalarToVectorCombine(SDNode *N, TargetLowering::DAGCombinerInfo &DCI,
+  auto MaskedLoad = dyn_cast<MaskedLoadSDNode>(Op0->getOperand(0));
+  if (!MaskedLoad || !MaskedLoad->hasNUsesOfValue(NumParts, 0) ||
+      MaskedLoad->getExtensionType() != ISD::NON_EXTLOAD ||
+      MaskedLoad->getAddressingMode() != ISD::UNINDEXED ||
```
`isNormalLoad` doesn't work for masked loads, so I've added a new `isNormalMaskedLoad`

david-arm commented:
Rebased and ran "make check-all", all seems fine.
